src/schedule.py

```diff
@@ -310,6 +310,67 @@ def log_register_use(computation_graph, step, hw_element_counts, execution_time)
     keys.sort()
     in_use_sorted = {i: in_use[i] for i in keys}
     write_df(in_use_sorted, hw_element_counts, execution_time, step)
+
+import networkx as nx
+from itertools import combinations
+
+def check_valid_hw(computation_graph, hw_netlist):
+    """
+    Checks if for every pair of operators, that isn't already bidirectionally connected, 
+    there exists at least one register that both operators are bidirectionally connected to.
+    """
+    hw_graph = nx.DiGraph()
+
+    for node_id, attrs in hw_netlist['nodes'].items():
+        hw_graph.add_node(node_id, **attrs)
+
+    for edge in hw_netlist['edges']:
+        hw_graph.add_edge(*edge)
+
+    register_nodes = [node for node, data in hw_graph.nodes(data=True) if data.get('type') == 'memory' and data.get('function') == 'Regs']
+
+    operator_nodes = [node for node, data in hw_graph.nodes(data=True) if data.get('type') == 'pe']
+
+    for op1, op2 in combinations(operator_nodes, 2):
+        if hw_graph.has_edge(op1, op2) and hw_graph.has_edge(op2, op1):
+            continue
```

Comment:
ooh nice touch. Ok with this we can ignore what I said in the next comment. I'm going to go ahead and merge.